### PR TITLE
refactor: use nix to manage development dependencies

### PR DESCRIPTION
**Pull request by cpcloud**
This is a PR to change the way local development is done in Ibis (while preserving as much backwards compatibility with conda and non-conda/non-nix workflows as possible!), accomplishing the following things:

The main change here is the addition of a development environment that is self-contained and derived from a single source of truth (`pyproject.toml`), using `nix`.

See `docs/web/contribute.md` for more details.

Note that **both conda and setuptools-based development workflows are still supported as first class citizens**.

DONE:
1. ~A move to [`poetry`](https://python-poetry.org) for dependency management.~ done in #2937 
   It is still possible and supported to use setup.py and setuptools if you like, this functionality is checked in CI.
1. ~More extensive CI testing, with minimal additional CI time used. It looks
   like roughly on average 3 minutes of additional CI time occur, with more of
   the library being tested on Windows in particular~ done in #2937 
1. ~Automatic generation of a conda recipe for every release, uploaded to GitHub
   releases (this is the remaining 1% that isn't automated but this can
   be automated by pushing a feedstock pull request directly from CI)~ I plan to do this in conda-forge directly, using their integrated bot automation
1. ~Automated Dependabot updates for github-actions and poetry dependencies~ this is done using Renovate in #3053
1. ~Generating a setup.py file from the project's `pyproject.toml` file~ ~This is now just checked that it doesn't need to be regenerated, since we want to ensure this is tested in CI where relevant.~ done in #3054, necessary for Renovate dependency update PRs
1. ~Automatic generation of a conda environment file, uploaded to GitHub releases, useful for developers using conda~ We have automatically generated lockfiles for linux, macos, and windows, for all three versions of Python currently supported (3.7, 3.8, 3.9). These live under `conda-lock` and can be used to set up a development environment very quickly since there's no solve step required. Done #3077, #3080, #3083, #3087

Follow ups:
1. `workflow_dispatch` (click-to-run) GitHub Actions workflow that cuts a release to PyPI and GitHub Releases.
1. Automatic license updates once a year or on-demand. This is optional, but it's another thing we can automate at some point.
1. Automatic nix dependency updates via PR submission every six hours or on-demand. This has to be done in a follow up.

Happy to address any concerns here, let's automate everything!

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
can you add a README.md explaining how the poetry2conda is used?

how are the min version set now for packages (e.g. where). 

also is the release process documented somewhere? (even if fully automated)

File: ibis/__init__.py
Comment (cpcloud) [reply ibis/__init__.py:50]: The bump happens automatically. I will add a commit that triggers a major version bump once the review is done.

File: docs/source/conf.py
Comment (cpcloud) [reply docs/source/conf.py:76]: Bump is automatic, I'll add a commit that triggers one after review is done.

File: docs/web/contribute.md
Comment (cpcloud) [reply docs/web/contribute.md:137]: Yup, here's a screenshot of running `cz` (which replaces `git commit`) that shows a subset of what's possible:

image

Let's say I'm committing a feature, so I select `feat`. I'd get this prompt now:

image

Here you can put whatever component or filename you want, let's say I'm editing sql generation, I'd then get prompt to add a commit title:

image

You then get asked for a longer description which is optional unless there's a breaking change: 
image

... breaking changes are the next question: 
image

Finally you're asked about open issues the commit affects:
image

The commit conventions are taken from the Angular community: https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines

Comment (jreback) [reply docs/web/contribute.md:137]: can you add this to the docs :-> (screen shots and everything). as this is a new workflow

File: .pre-commit-config.yaml
Comment (cpcloud) [reply .pre-commit-config.yaml:2]: This is automatically generated upon running `nix-shell -A python39`, I'll add some documentation to this effect.

**Comment by jreback** [comment]
cc @datapythonista

**Comment by cpcloud** [comment]
@jreback 

> can you add a README.md explaining how the poetry2conda is used?

Yup. In short it's used to generate an environment file in CI.

> how are the min version set now for packages (e.g. where).

The minimum versions for dependencies are set in `pyproject.toml`

> also is the release process documented somewhere? (even if fully automated)

Light documentation of it exists in `docs/web/contributing..md`, but I'll spell it out in more detail there.

File: docs/web/contribute.md
Comment (jreback) [reply docs/web/contribute.md:137]: i c.

ok then prob a good idea to show this workflow AND show that link above for the allowed messages.

Comment (cpcloud) [reply docs/web/contribute.md:137]: Yup, will add to the docs for sure.

File: ibis/backends/pandas/trace.py
Comment (icexelloss) [inline ibis/backends/pandas/trace.py:127]: what are these changes?

Comment (cpcloud) [reply ibis/backends/pandas/trace.py:127]: This fixes a bug in testing, where there's an ordering dependency that assumes that `import pandas` happens before executing certain operations. In backends that use the native pandas backend implicitly (like computing constants) this code path can be hit _without_ having run `import pandas` and so the configuration option is never registered and thus fails. This can be reproduced by running tests with `pytest-xdist` and setting `--numprocesses` to something larger than 1.

File: docs/web/contribute.md
Comment (icexelloss) [reply docs/web/contribute.md:137]: If this is verified in CI then I would say it's necessary to use `cz` (which is fine), otherwise I can see the commit message doesn't pass CI easily.

File: ci/deps/pyspark-min.yml
Comment (icexelloss) [inline ci/deps/pyspark-min.yml:2]: How are these dependency handled now?

File: .prettierrc.toml
Comment (cpcloud) [reply .prettierrc.toml:1]: formatting toml, yaml, and json files in the repo

File: .pre-commit-config.yaml
Comment (icexelloss) [inline .pre-commit-config.yaml:1]: which file does pre-commit configs now?

Comment (cpcloud) [reply .pre-commit-config.yaml:1]: The configuration of the auto gen'd file is in `pre-commit.nix`.

File: ibis/backends/pandas/tests/test_client.py
Comment (icexelloss) [inline ibis/backends/pandas/tests/test_client.py:13]: Why this change?

Comment (cpcloud) [reply ibis/backends/pandas/tests/test_client.py:13]: For the same reason as the other `import pandas` addition: configuration options aren't registered unless `ibis.pandas` is imported, and when tests are run in a non-deterministic order that occasionally means `ibis.pandas` isn't executed before this fixture is used to execute things.

**Comment by icexelloss** [comment]
@cpcloud left some comments, in general +1 to standardize/automatic things so thanks for this! 

My main concern is that I haven't used many of the tools here (I think some other folks as well), so would be really helpful to write dev docs describing how the new CI/dev/release works.

File: ci/deps/pyspark-min.yml
Comment (cpcloud) [reply ci/deps/pyspark-min.yml:2]: There's no explicit handling of them. Happy to add it back if necessary, but pyspark tests are all passing. Is there a specific configuration that isn't passing? If so, can you describe it? Happy to add more changes to address if needed.

**Comment by cpcloud** [comment]
> @cpcloud left some comments, in general +1 to standardize/automatic things so thanks for this!
> 
> My main concern is that I haven't used many of the tools here (I think some other folks as well), so would be really helpful to write dev docs describing how the new CI/dev/release works.

Heck yeah! @jreback has requested more docs as well. I think I'll also add a screencast to really help clarify how these things are used.

File: ci/deps/pyspark-min.yml
Comment (cpcloud) [reply ci/deps/pyspark-min.yml:2]: FYI, this patch includes an upgrade of pyspark from 2->3, is that problematic? Happy to downgrade if it is.

File: ibis/backends/pandas/trace.py
Comment (icexelloss) [reply ibis/backends/pandas/trace.py:127]: Actually, what is `ibis.pandas` namespace now? I think pandas backend code is move to `ibis.backends.pandas` already

Comment (cpcloud) [reply ibis/backends/pandas/trace.py:127]: `ibis.pandas` is a dynamic module attribute taken from the notion of an `entry_point`: https://github.com/ibis-project/ibis/blob/master/ibis/__init__.py#L56

File: ibis/backends/tests/test_client.py
Comment (cpcloud) [inline ibis/backends/tests/test_client.py:220]: These changes fix a race condition when running tests in parallel. Without this multiple clients try to remove the `temp_to_table` independently.

File: ibis/tests/test_version.py
Comment (cpcloud) [inline ibis/tests/test_version.py:15]: I think this test is completely useless, but I'm going to revert the changes here as they are unnecessary.

File: ibis/backends/pyspark/tests/conftest.py
Comment (cpcloud) [inline ibis/backends/pyspark/tests/conftest.py:202]: This is another fix that is necessary to get pyspark tests passing. It's not clear to me whether the current CI is exercising this code path.

File: docs/web/release_notes.md
Comment (cpcloud) [inline docs/web/release_notes.md:3]: FYI this is where new release notes based on commit logs will start to be automatically populated upon release.

File: ibis/tests/test_version.py
Comment (cpcloud) [reply ibis/tests/test_version.py]: Reverted.

**Comment by cpcloud** [comment]
@jreback @icexelloss I've added a bunch of docs on the dev and release process. Please review at your earliest convenience!

File: ibis/backends/tests/test_client.py
Comment (cpcloud) [reply ibis/backends/tests/test_client.py:220]: I'll move this to another patch.

File: ibis/backends/pyspark/tests/conftest.py
Comment (cpcloud) [reply ibis/backends/pyspark/tests/conftest.py:202]: Moving to another patch

File: ibis/backends/tests/conftest.py
Comment (cpcloud) [inline ibis/backends/tests/conftest.py:63]: Moving to a separate patch.

Comment (cpcloud) [reply ibis/backends/tests/conftest.py:63]: #2927

File: ibis/backends/pyspark/tests/conftest.py
Comment (cpcloud) [reply ibis/backends/pyspark/tests/conftest.py:202]: #2926

File: ibis/backends/tests/test_client.py
Comment (cpcloud) [reply ibis/backends/tests/test_client.py:220]: #2925

**Comment by gerrymanoim** [comment]
A nice addition here might be to  also add a `devcontainer.json` file that bootstraps a dev env, both in vscode and github codespaces. 

https://docs.github.com/en/codespaces/customizing-your-codespace/configuring-codespaces-for-your-project

**Comment by cpcloud** [comment]
> A nice addition here might be to also add a `devcontainer.json` file that bootstraps a dev env, both in vscode and github codespaces.
> 
> https://docs.github.com/en/codespaces/customizing-your-codespace/configuring-codespaces-for-your-project

Interesting, let me look into that.

**Comment by cpcloud** [comment]
I was able to get a environment working in codespaces.

@gerrymanoim would you be ok with that in a follow up? The branch is up at `devcontainer` on my fork if you want to try it.

**Comment by datapythonista** [comment]
Do you think you can split this PR into multiple parts, and get this merged incrementally? Changes here are huge, and will be very inefficient to have to review everything at each iteration. And there are many things that are somehow independent to the final stage, like adding a pyproject.toml, formatting changes to CI files...

**Comment by cpcloud** [comment]
> Do you think you can split this PR into multiple parts, and get this merged incrementally? Changes here are huge, and will be very inefficient to have to review everything at each iteration. And there are many things that are somehow independent to the final stage, like adding a pyproject.toml, formatting changes to CI files...

Some of it, yeah. I can revert the formatting changes and add those to a follow up. The refactoring to use nix and the refactoring of `main.yml`, and all the file deletions need to be done together though, to prevent breakage. Many of the files "changed" are files that are being removed because they are no longer necessary.

**Comment by cpcloud** [comment]
@datapythonista I've pared down the PR quite a bit, and have removed all of the formatting changes and a few workflows that I wlll add in separate PRs once this lands.

**Review by datapythonista — COMMENTED** [review]
This looks very interesting, but I need to read about nix, poetry... as I haven't used them yet, before I can provide further feedback.

This also looks scary. We had lots of problems with the CI in the past. I'm not sure, but it feels we could still do this more incrementally, and minimize risks. For example, can commitlint be implemented earlier? Feels like we can add a job for it in the current CI, without the rest of stuff. Even if the automatic generation of the release notes is added with nix, feels like having that earlier would give us feedback in a more controlled way, and splitting this out of this PR, would make both PR diffs smaller and easier to review.

I guess you have it under control, but I'm wondering about how the commit linting works with GitHub commit squash. If it doesn't work as expected, having it merged in a HUGE pull request like this, with lots of other things possibly failing, can be very disruptive.

And I think there are several other cases where we can merge things early. Like, can we move the isort config from `setup.cfg` to `pyproject.toml` before moving to nix? Can we start using poetry before nix too?...

If I'm wrong, and we need to merge all this together, I'd prefer to merge this after the release of Ibis 2.0. I'm worried this can significantly delay the release, which would be good to have out asap.

File: pyproject.toml
Comment (cpcloud) [reply pyproject.toml:164]: Roger that, I'll remove them.

**Comment by cpcloud** [comment]
@datapythonista 

> For example, can commitlint be implemented earlier? Feels like we can add a job for it in the current CI, without the rest of stuff.

Yes, but I'm not how we'd bring the dependency in without nix. We probably can, but it doesn't seem worth the trouble just to delete the code later.

> I guess you have it under control, but I'm wondering about how the commit linting works with GitHub commit squash. If it doesn't work as expected, having it merged in a HUGE pull request like this, with lots of other things possibly failing, can be very disruptive.

The Python implementation of semantic release does not support squash merging.

Honestly, squash versus rebase merging seems like bikeshedding if you have a commit structure being enforced by a tool. IMO the automation is totally worth not having a single commit for every PR (I do understand the value of this, but it's overly restrictive IMO if you lose the ability to remove toil from maintainers' lives). None of us really have to think too hard about releases anymore, and versioning becomes unsentimental and meaningful.

> And I think there are several other cases where we can merge things early. Like, can we move the isort config from setup.cfg to pyproject.toml before moving to nix? 

Sure, that's doable. I will see if there's anything else like this and open one or more PRs.

> Can we start using poetry before nix too?

We can, but it's not clear to that that will reduce the number of changes by a significant amount. I will look into whether that can be split.

> If I'm wrong, and we need to merge all this together, I'd prefer to merge this after the release of Ibis 2.0. I'm worried this can significantly delay the release, which would be good to have out asap.

That's fair. FWIW, after these changes are landed (in a sequence of PRs or otherwise), releases will happen automatically, so modulo the rare CI flakiness, we won't have to worry about release delays ever again.

**Comment by kszucs** [comment]
Splitting it into two PRs would make it easier for us to review. I definitely like the switch to poetry, especially with the conversion scripts to setup.py and conda recipe, though I'm not so certain about nix.

**Comment by gerrymanoim** [comment]
> @gerrymanoim would you be ok with that in a follow up? The branch is up at `devcontainer` on my fork if you want to try it.

@cpcloud Definitely fine as a follow up, mostly a nice to have that I wanted to get in at some point.

**Comment by cpcloud** [comment]
We should merge the poetry PR first, which will greatly reduce the size of this PR.

**Comment by cpcloud** [comment]
Another thing to remember here: developers are not required to use `cz` or any of the tooling here, conda, setuptools, and poetry dev workflows are all supported.

The new tooling will run in CI and validate that commit messages have the right structure for semantic release.

**Comment by cpcloud** [comment]
The pre-commit check will continue to fail until this is merged, after which I can turn it off since CI is checking pre-commit after this PR.

File: docs/web/contribute.md
Comment (jreback) [inline docs/web/contribute.md:95]: since this is not specific to nix, can you be more generic here

File: docs/web/contribute.md
Comment (jreback) [inline docs/web/contribute.md:152]: is the pre-commit still available? i think you removed the config?

File: docs/web/contribute.md
Comment (jreback) [inline docs/web/contribute.md:145]: is there a doc link on cz?

File: docs/web/contribute.md
Comment (jreback) [inline docs/web/contribute.md:180]: umm what?

**Review by jreback — CHANGES_REQUESTED** [review]
umm why all the references to 'breaking changes' here? why is this in any way controlled by the commits that someone pushes? how could they possibly know if something is a breaking change or not?

sure the maintainers can know / understand this, but these are general instructions.

**Comment by cpcloud** [comment]
@jreback

> umm why all the references to 'breaking changes' here?

It seems like you might not have had a chance to take a look at [`semantic-release`](https://github.com/semantic-release/semantic-release)? In particular [this section](https://github.com/semantic-release/semantic-release#how-does-it-work) answers this question.

> why is this in any way controlled by the commits that someone pushes?

The answer to this question is also in the above links, but I'll give a brief summary.

The point of semantic release is to entirely automate the drudgery of release. This is done by giving semantics to commit messages, e.g., `feat` is a feature (causing a minor version bump), `bug` is a bug causing a patch version bump. The structure is enforced by `cz` (optional to use) and enforced by the `commitlint` check in CI (not optional).

I think you may be missing the fact that these version determinations are **batched at release time**: there won't be a new release _per commit_, that would be excessive.

Instead, the commits between the last release and the current commit are examined for `feat`, `bug` etc. If `feat` is found, it's a minor bump (from say `2.0.0` to `2.1.0`), if `bug` is found it's a patch bump (from say `2.0.0` to `2.0.1`). There are a few other rules (notably any message with `BREAKING CHANGE: blah blah blah` triggering a major version bump) but what I just wrote captures most of the important information.

All of this is covered in the semantic-release repo, please take a look at that.

Also note that this PR doesn't implement the release automation. I explicitly state that it'll be done in a follow up in the PR description. Did you read that? Please do so if you haven't.

> how could they possibly know if something is a breaking change or not?

Not sure who `they` is here, but if `they` means "a developer", then those are the only people who can know what breaks and what doesn't before a release.

If by `they` you mean "commit messages", then again, that's covered in the semantic-release repo link above.

File: docs/web/contribute.md
Comment (cpcloud) [reply docs/web/contribute.md:180]: Please read this https://github.com/semantic-release/semantic-release.

It would also be helpful to articulate a question, since just "what?" does give me a whole lot to go on 🤣

File: docs/web/contribute.md
Comment (cpcloud) [reply docs/web/contribute.md:145]: I don't know what you're asking. Do you want me to put a link to the `cz` repository/documentation?

File: docs/web/contribute.md
Comment (cpcloud) [reply docs/web/contribute.md:152]: Yes. I've already explained this elsewhere in the PR.

File: docs/web/contribute.md
Comment (cpcloud) [reply docs/web/contribute.md:95]: Can you be more descriptive in what you're asking for?

File: .github/workflows/ibis-backends.yml
Comment (cpcloud) [reply .github/workflows/ibis-backends.yml:385]: I guess I can split this out, though I'd rather just merge this together since the PR has been open for a very long time.

**Comment by jreback** [comment]
by they i mean - a person pushing commits

they cannot possibly know what is. breaking change or not

the maintainers could know but o am not sure how this helps

**Comment by cpcloud** [comment]
> by they i mean - a person pushing commits
> 
> they cannot possibly know what is. breaking change or not
> 
> the maintainers could know but o am not sure how this helps

Not sure what the objection is here. We don't have random people making commits. Every change goes through review. If a change needs a commit to indicate a breaking change, then it's up to maintainer to make sure that the commit has it. Typically this would be done by referring someone to the docs.

**Comment by jreback** [comment]
> If a change needs a commit to indicate a breaking change, then it's up to maintainer to make sure that the commit has it.

this seems like quite a lot of work - meaning the commits need to be exactly right 

having release notes as part of the change is way easier

**Comment by cpcloud** [comment]
> > If a change needs a commit to indicate a breaking change, then it's up to maintainer to make sure that the commit has it.
> 
> this seems like quite a lot of work - meaning the commits need to be exactly right 
> 
> having release notes as part of the change is way easier 
> 
> 

It's not a lot of work, and cz handles it. I've used this procedure at both past companies and in my own projects and whatever small overhead there is in making your commits more structured is easily made up by the ability to release with confidence, whenever we want. cz makes this process even easier because it'll prompt you, and construct the commit message correctly. Check out the prompts I show above, which you explicitly requested.

It might be useful to try it out, so that if you have specific objections I can address those.

We don't have that many breaking changes.

We do however need to release faster, and it needs to be a super boring and routine procedure that anyone can do with a single click.

I would like to move forward with this unless there specific actionable objections, in which case I will address them.

**Comment by jreback** [comment]
ok will look again tomorrow 

in any event would like to make the docs super clear about this

**Comment by jreback** [comment]
btw likely we are plaguaged by https://github.com/pypa/setuptools/issues/2941 (to mitigate can pin < 60)

**Comment by cpcloud** [comment]
> btw likely we are plaguaged by [pypa/setuptools#2941](https://github.com/pypa/setuptools/issues/2941) (to mitigate can pin < 60)

Thanks. For now it looks like this PR isn't affected.

@jreback On a different note, I've reworked the docs and made it even easier to use `cz` (it's now a `dev-dependency`). Let me know if there's anything else you'd like to me to add to the `contributing.md` docs.

**Comment by jreback** [comment]
kk will look shortly

File: docs/web/contribute.md
Comment (jreback) [inline docs/web/contribute.md:256]: followup: add a doc reference for this

File: docs/web/contribute.md
Comment (jreback) [inline docs/web/contribute.md:280]: these are in the conda files?

**Review by jreback — APPROVED** [review]
seems fine to merge on green.

skeptical about the auto release notes but that's in the future anyhow.

File: docs/web/contribute.md
Comment (cpcloud) [reply docs/web/contribute.md:256]: Happy to add this now.

File: docs/web/contribute.md
Comment (cpcloud) [reply docs/web/contribute.md:279]: Not explicitly, but they are both a `pip install`/`conda install` away.

**Comment by cpcloud** [comment]
Thanks.

I'm going to turn off the `pre-commit.ci` integration and move us to `rebase and merge` for merging. `commitlint` will take care of notification of non-conforming commit messages

**Comment by jreback** [comment]
thanks @cpcloud great

**Pull request merged**
